src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala

```diff
@@ -0,0 +1,138 @@
+  }
+
+  private def registerUDFs(session: SparkSession): Unit ={
+    session.sqlContext.udf.register("MeanAverage", FortisUdfFunctions.MeanAverage)
+    session.sqlContext.udf.register("SumMentions", FortisUdfFunctions.OptionalSummation)
+    session.sqlContext.udf.register("SentimentWeightedAvg", SentimentWeightedAvg)
+  }
+
+  /*Writes a Dstream of FortisEvents into the events table*/
+  private def writeEventRddToEventsTable(eventsRDD: RDD[FortisEvent], batchid: String): RDD[Event] = {
+    val fortisEventsDF = eventsRDD.map(CassandraEventSchema(_, batchid))
+    fortisEventsDF.saveToCassandra(KeyspaceName, TableEvent, writeConf = WriteConf(ifNotExists = true))
+
+    fortisEventsDF
+  }
+
+  private def fetchEventBatch(batchid: String, events: RDD[Event], session: SparkSession): Dataset[Event] = {
+    import session.sqlContext.implicits._
+
+    val addedEventsDF = session.sqlContext.read.format(CassandraFormat)
+      .options(Map("keyspace" -> KeyspaceName, "table" -> TableEventBatches))
+      .load()
+
+    addedEventsDF.createOrReplaceTempView(TableEventBatches)
+    val ds = session.sqlContext.sql(s"select eventid, pipelinekey " +
+      s"                     from $TableEventBatches " +
+      s"                     where batchid = '$batchid'")
+
+    val eventBatch = ds.as[EventBatchEntry].map(ev=>mergedEventIdentifier(ev)).collect.toSet
+    events.filter(ev=>eventBatch.contains(mergedEventIdentifier(ev))).toDF.as[Event]
+  }
+
+  private def mergedEventIdentifier(event: EventBase): String = s"${event.eventid}_${event.pipelinekey}"
+
+  private def writeEventBatchToEventTagTables(eventDS: Dataset[Event], session: SparkSession): Unit = {
+  import session.implicits._
+
+  val eventTopicsRddColumns = SomeColumns("topic", "pipelinekey", "externalsourceid", "insertiontime", "eventids" append)
+  val eventPlacesRddColumns = SomeColumns("conjunctiontopic1", "conjunctiontopic2","conjunctiontopic3", "insertiontime", "centroidlat", "centroidlon", "pipelinekey", "placeid", "externalsourceid", "eventids" append)
+  //Need to convert the DS to a RDD as C* column append isnt support for Dataframes
+  saveRddToCassandra[EventTopics](KeyspaceName, TableEventTopics, eventDS.flatMap(CassandraEventTopicSchema(_)).rdd, eventTopicsRddColumns)
+  saveRddToCassandra[EventPlaces](KeyspaceName, TableEventPlaces, eventDS.flatMap(CassandraEventPlacesSchema(_)).rdd, eventPlacesRddColumns)
+}
+
+private def saveRddToCassandra[T](keyspace: String, table: String, rdd: RDD[T], columns: ColumnSelector, attempt: Int = 0)(
+                                 // implicit parameters as a separate list!
+                                 implicit rwf: RowWriterFactory[T],
+                                 columnMapper: ColumnMapper[T]
+                               ): Unit = {
+Try(rdd.saveToCassandra(keyspace, table, columns)) match {
+  case Success(_) => None
+  case Failure(ex) =>
+    ex.printStackTrace
+    Thread.sleep(CassandraRetryIntervalMS)
```

Comment:
I wonder if you want this as an exponential backoff instead of a linear one: `CassandraRetryIntervalMS * attempt` if we initialize the attempt to 1 instead of zero on line 116 below.